[Project1/채수혁] Notion 클로닝 프로젝트

ienrum opened the pull request:
## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
![example](https://github.com/prgrms-fe-devcourse/FEDC5-5_Project_Notion_VanillaJS/assets/137467530/c928ee8f-d27c-412c-b4ae-a944e797efb0)


## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
- [x]  api 를 통해 documents 렌더링
    - [x]  왼쪽 사이드바에 트리형태 렌더링
    - [x]  root documet 클릭시 편집기에 렌더링
    - [x]  플러스버튼 클릭시 생성과 편집기에 렌더링
- [x]  api 를 활용하여 서버 저장
    - [x]  지속적인 서버 api 저장
    - [x]  로컬 스토리지와 연동
        - [x]  실패하면 ou 취소
- [x]  spa 형태로 구현
    - [x]  편집기 렌더링시 /documets/{documetId} 로 불러와 구현

### 추가 구현
- [x]  textarea ⇒ rich editor
- [x]  편집기 최하단에 하위 document 를 링크하도록 렌더링
- [x]  편집기 title 이 이미있는 title 이라면 그쪽으로 링크 걸기
<!--  ## ✅ 피드백 반영사항  지난 코드리뷰에서 고친 사항을 적어주세요. 재PR 시에만 사용해 주세요! (재PR 아닌 경우 삭제) -->

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
- editor.js 컴포넌트에서 this.setLinkPosts 함수를 이 컴포넌트 내에서만 쓰일것 같아 내부적으로 작성했는데 따로 파일로 빼두는게 맞을까요??
- css 처리를 나중에 하다보니 $target 부분을 상위 컴포넌트에서 위임받지않고 만들어 쓰게되었는데, 보통 이런문제는 css 랑 동시에 만들면서 처리하시나요?? 아니면 미리 방지할수있는 부분이있을까요?
- 함수형 코딩이란 책을 읽으면서 함수형으로 짜보려고 노력했는데, 전혀 그렇게 된것같지 않습니다. 처음에만 이런건가요? 아니면 원래 함수형 코딩을 적용하기가 까다롭나요? 그러면 어떻게해야하나요?
- 커밋 단위가 적절히 나눠져있을까요?? 기능 하나당 어느정도 범위의 기능이어야하나요??

ienrum committed:
✨ : index.html 추가

ienrum committed:
✨ : 최상위 컴포넌트 main.js 추가

fetch 인터페이스 함수들로 하위 컴포넌트들에게 initial 데이터 주기

ienrum committed:
✨ : header 컴포넌트 추가

main.js 에서 받은 텍스트로 렌더링

ienrum committed:
✨ : main 하위 컴포넌트인 app.js 추가

main 에서 받은 initial 데이터를 app 의 하위컴포넌트 들에게 전달

ienrum committed:
✨ : editor.js 컴포넌트 추가

input, textarea 로 입력을 받아 fetch 의 put 인터페이스로 document 정보 수정, 이때 sidebar 를 리렌더링 시킴

ienrum committed:
✨ : sideBar.js 컴포넌트 추가

이벤트 위임으로 button , li 각각에 이벤트 추가 (fetch 인터페이스로 post , 조회 기능 수행)
완료시 editor 에 id 값 전달.

documentTreeToHTML 함수를 활용하여 트리구조인 documents 를 렌더링

ienrum committed:
✨ : dataManager.js 유틸 파일 추가

storage.js 와 fetchData.js 로 얻은 데이터를 조작하는 함수들 구현 파일

ienrum committed:
✨ : fetchData.js 파일 추가

fetch 함수를 활용한 인터페이스 함수들 구현 및 fetch 에 필요한 데이터 설정.

ienrum committed:
✨ : storage.js 파일 추가

localStorage 를 이용하기위한 인터페이스 함수들 추가

ienrum committed:
💄 : index.css 파일 추가

flex 로 헤더와 메인을 나눔.
다시 메인을 플렉스로 editor 와 sidebar 를 나눔

화면이 작아지면 글자크기도 작아짐

ienrum committed:
🐛 : getIDs 가 없을경우 예외처리 수정

ienrum committed:
♻️ : fetch -> state 로 직접 상태 변경하도록 수정, 함수명 변경

fetch 로 상태를 간접적으로 변경하는것에서 직접 상태변경을 콜백함수로 수행 하도록 변경

ienrum committed:
✨ : 특정 postid 인 post 가 localstorage 에 저장되었다면 가져오는 기능 추가

ienrum committed:
♿️ : post 를 새로생성시, 리렌더링 되도록 변경

ienrum committed:
🔇 : 디버깅코드 삭제

ienrum committed:
✨ : storage.js 에 기존 함수들을 인터페이스화한 함수들 추가

ienrum committed:
✨ : 데이터를 명시적으로 업데이트와 데이터를 명시적으로 추가하는 코드 추가

ienrum committed:
♿️ ♻️ : 낙관적 업데이트 적용 및 컴포넌트 분할

PostEdit.js 를 새로 만들어 Editor.js 역할을 분할

ienrum committed:
✨ : history api 로 spa 형태 구현

spa 404 에러 처리 및 /documents/${documentsid} 형태로 spa 구현

ienrum committed:
🐛 : 에디터에 falsy 값 처리

ienrum committed:
🚚 : router.js 를 utils 폴더로 옮김

ienrum committed:
✨ : 리치 에디터 추가

RichView 컴포넌트를 추가하여 리치 에디터 뷰 추가

mdToHtml 함수로 렌더링

ienrum committed:
✨ : 라우터 로 편집기 최하단에 하위 document 를 링크 추가

footer.js 와 router.js  기능 추가

footer js 에 하위 documets link 추가 하고 router.js 의 push 로 클릭시 이벤트 발생시키도록 함.

ienrum committed:
✨ : 콘텐츠 내에 존재하는 포스트의 제목이있다면 링크처리 기능 추가

에디터에 링크로 변환하는 함수를 추가함
rich view 에서 변환된 md 문법을 a 태그로 변환시켜 rich view 에 출력

ienrum committed:
💄 : ui를 위한 html 위치 변경

ienrum committed:
💄 : css 수정

ienrum committed:
🐛 : 정규표현식 을 활용한 link 달기 기능 수정

\b 를 활용할때 ``로 감싼다면 \\b 여야한다.

ienrum committed:
✨ : content 가 비어있을시 fetch delete 로 삭제 기능 추가

ienrum committed:
✨ : 뒤로가기, 앞으로가기 렌더링 추가

ienrum committed:
♻️ : utils 함수 리팩토링

주석 및 코드 변수 리팩토링 했습니다.

ienrum committed:
♻️ : 컴포넌트 리팩토링

주석 및 변수명 변경
자잘한 버그 수정

ienrum committed:
♿️ : 첫 화면이 빈화면이 아니라 첫번째 document 이도록 수정

ienrum committed:
📝: gif 추가가

mojoblue7 commented on src/main.js:
```diff
@@ -0,0 +1,17 @@
+import { getData, getDatas } from "./utils/fetchData.js";
+import { getIDs } from "./utils/dataManager.js";
+
+const initialSideBarState = await getDatas();
+const initialId = getIDs(initialSideBarState)[0];
+
+let initialEditorState = { id: -1, title: "", content: "" };
+
+if (typeof initialId === "number") {
+  initialEditorState = await getData(initialId);
+}
```
await의 경우 async function에 내부에서 사용해야합니다. 
[https://ko.javascript.info/async-await](https://ko.javascript.info/async-await)
위 내용을 확인해보고 적용해보면 좋을 것 같습니다 😃

ienrum replied:
넵! 감사합니다.. 참고자료 살펴보겠습니다.!

mojoblue7 reviewed (commented):
프로젝트 하시느라 고생많으셨습니다 :)

부가적으로 삭제 기능이 있었으면 좋겠다고 생각했습니다. 또한 새 문서를 추가했을때, 타이틀이 빈칸이어서 메뉴에 있는 타이틀 높이가 짧아 누르기 힘들었던 점도 있어보였습니다.

Q1 ) editor.js 컴포넌트에서 this.setLinkPosts 함수를 이 컴포넌트 내에서만 쓰일것 같아 내부적으로 작성했는데 따로 파일로 빼두는게 맞을까요??
- 에디터 내부에서 사용하는 것이라면 굳이 뺄 필요는 없어보입니다. 기능 상으로 확인했을때도 빼지 않아도 된다고 생각합니다.

Q2 ) css 처리를 나중에 하다보니 $target 부분을 상위 컴포넌트에서 위임받지않고 만들어 쓰게되었는데, 보통 이런문제는 css 랑 동시에 만들면서 처리하시나요?? 아니면 미리 방지할수있는 부분이있을까요?
- 질문의 요지와 어느 부분에서의 문제인지 잘 인지하지 못했는데요. 커피챗때 코드를 확인하면서 다시 대화 나눠보시면 좋을 듯 합니다 😃

Q3 ) 함수형 코딩이란 책을 읽으면서 함수형으로 짜보려고 노력했는데, 전혀 그렇게 된것같지 않습니다. 처음에만 이런건가요? 아니면 원래 함수형 코딩을 적용하기가 까다롭나요? 그러면 어떻게해야하나요?
- 제 의견은 어느 정도 잘 맞춰 짜신 것 같은데요, 개인적으로는 함수형 프로그래밍에 대한 패러다임에 매몰되지는 않으셨으면 좋겠습니다. 함수형 프로그래밍의 경우는 패러다임 적용을 위한 라이브러리가 필요한 경우도 있고, 작성자가 아니면 가독성이 어려워지는 경우도 있습니다. 그래서 상황에 맞게 적용하는 것이 좋은 방법이라고 생각합니다. 참고로 제 의견은 정답이 아니기에 여러 글들을 읽어보시며 깊은 고찰이 부분이라고 말씀드리고 싶습니다 😃

Q4 ) 커밋 단위가 적절히 나눠져있을까요?? 기능 하나당 어느정도 범위의 기능이어야하나요??
- 정답이 없는 부분이어서 의견을 드립니다. 파일의 추가 단위나 기능 단위로 세세하게 작업하신 케이스 같은데요, 저 같은 경우에는 기능 단위 당으로 커밋합니다. 다른 사람이 보았을 때 어렵지 않은 단위이면 좋습니다. 프론트의 경우는 컴포넌트 생성이 단위가 될 수는 있을 것 같습니다.

bel1c10ud commented on src/utils/mdToHtml.js:
```diff
@@ -0,0 +1,46 @@
+export function mdToHtml(mdText) {
+  let prevLine = "";
+  mdText = mdText.split("\n").map((line) => {
+    let result = line.replace(H1_REGEX, makeHtmlTag("h1", "$1", style));
+    result = result.replace(H2_REGEX, makeHtmlTag("h2", "$1", style));
+    result = result.replace(H3_REGEX, makeHtmlTag("h3", "$1", style));
+    result = result.replace(BOLD_REGEX, makeHtmlTag("b", "$1", style));
+    result = result.replace(ITALIC_REGEX, makeHtmlTag("i", "$1", style));
+    result = result.replace(LIST_REGEX, makeHtmlTag("li", "$1", style));
+    result = result.replace(LINK_REGEX, makeHtmlTag("a", "$1", "href=$2"));
```
```js
const result = line.replace(H1_REGEX, makeHtmlTag("h1", "$1", style))
    .replace(H2_REGEX, makeHtmlTag("h2", "$1", style))
    .replace(H3_REGEX, makeHtmlTag("h3", "$1", style))
    .replace(BOLD_REGEX, makeHtmlTag("b", "$1", style))
    .replace(ITALIC_REGEX, makeHtmlTag("i", "$1", style))
    .replace(LIST_REGEX, makeHtmlTag("li", "$1", style))
    .replace(LINK_REGEX, makeHtmlTag("a", "$1", "href=$2"));
```
이렇게 작성하면 const로 바꿀 수 있을 것 같아요🙂

ienrum replied:
오 감사합니다!!

bel1c10ud commented on src/utils/mdToHtml.js:
```diff
@@ -0,0 +1,46 @@
+
+// html tag 를 만들어주는 함수입니다.
+function makeHtmlTag(tagName, content, feature, listCount) {
+  return `<${tagName} ${feature}>${content}</${tagName}>`;
```
🤔 PR 포인트에서 언급하신 부분이 이 부분이 맞다면 
저는 인라인 스타일 말고 css를 같이 작성해서 css 클래스를 지정해서 스타일을 적용했어요 🙂

ienrum replied:
아 이건 제가 글로 표현을 잘 못해서 발생한문제같아요ㅜ, 코드를 작성하다보니 처음에 잡았던 구조와 달리 html 레이아웃 구조가 바뀌게되면서 문제가 발생했는데요, css 를 기존 레이아웃 구조에 맞게 작성했다가 html 구조가바뀌니까 망가지게 되어서, 이를 어떻게 해결해야하는지 궁금했습니다..

bel1c10ud reviewed (commented):
전반적으로 꼼꼼하게 함수를 잘 분리하신 것 같아요 👍
고생 많으셨습니다.

kimyou1102 commented on src/main.js:
```diff
@@ -0,0 +1,17 @@
+import { getIDs } from "./utils/dataManager.js";
+
+const initialSideBarState = await getDatas();
+const initialId = getIDs(initialSideBarState)[0];
```
문서들 중 첫번째 문서의 id값만 필요하다면 initialSideBarState[0].id도 좋을 거 같습니다🙂

kimyou1102 commented on src/utils/dataManager.js:
```diff
@@ -0,0 +1,113 @@
+
+// api get 요청으로 받은 트리구조 Documents 데이터를 callback 함수를 통해 수정합니다.
+// dfs
+export function updateDocumentToTree(documents, id, data, callback) {
```
callback함수보다 명확하게 알 수 있게끔 documentItemChange 같은 변수명은 어떨까요?😀

ienrum replied:
앗 감사합니다.! Callback 이라고 명시해버렸네요..

kimyou1102 commented on src/utils/mdToHtml.js:
```diff
@@ -0,0 +1,46 @@
+const LINK_REGEX = /\[([^\]]+)\]\(([^)]+)\)/g;
+
+const CHECKBOX_REGEX = /^\[(x| )\] (.*)/g;
+const NEW_LINE_REGEX = /\n/g;
```
사용하지 않는 코드는 제거해주셔도 좋을 거 같습니다🙂

ienrum replied:
넵 다시 정리하겠습니다! 감사합니다!

kimyou1102 reviewed (commented):
전체적으로 주석덕분에 이해하기 쉬웠고 함수를 잘게 분리하신 것을 보며 많이 배워갑니다! 고생하셨습니다:)

ienrum commented on src/components/Editor.js:
```diff
@@ -0,0 +1,120 @@
+    // 찾은 타이틀 중에 중복되는 타이틀을 제거하고 링크에 포함된 타이틀을 제거합니다.
+    // ex) [post1](/documents/post1) => post1
+    const tempLinkingPosts = foundedTitleInContent.reduce((acc, cur) => {
+      if (cur.title === "documents" || cur.title === "") return acc;
```
한번에 이해하기 쉬운것 같습니다. 감사합니다!

ienrum commented on src/components/Editor.js:
```diff
@@ -0,0 +1,120 @@
+import { setEditingPostData, removeEditingPostData } from "../utils/storage.js";
+import { debounce } from "../utils/debounce.js";
+
+export default function Editor({ $target, initalState, setSideBarState }) {
+  const $div = document.createElement("div");
+  $div.id = "editor";
```
아하 감사합니다..!

ienrum commented on src/components/Editor.js:
```diff
@@ -0,0 +1,120 @@
+  };
+
+  // debounce 를 사용하여 1초마다 서버에 데이터를 전달합니다.
+  this.debounce = debounce(() => {
+    const body = {
+      title: this.state.title,
+      content: this.state.content,
+    };
+    putData(this.state.id, body).then((data) => {
+      // sideBar 컴포넌트에 상태를 전달합니다.
+      setSideBarState(data);
+      // 서버에 데이터를 전달한 후에는 localStorage 에서 해당 데이터를 삭제합니다.
+      removeEditingPostData(this.state.id);
+    });
+  }, 1000);
```
넵 여기서 await 를 사용할수가 있었군요 감사합니다!

ienrum committed:
♻️ : pr 피드백 적용

then -> async/await
최상위 파일에서 즉시실행함수로감싸 async/await
editor, postview 매개변수로 id 받아 설정
postEdit 의 id 값이 같다면 setState 취소
제
더 나은 변수명및 함수명
함수 체이닝으로 변경
쓰이지않는 코드 삭제
